### DISCUSSION TITLE
ALF v 7.2.1 Failed to open book

### DISCUSSION THREAD

**Discussion by Chexote** [apprenticeharper/DeDRM_tools#2022] [Q&A]
SUMMARY:
I installed a new version of Calibre. Installed Alf.  Tried to import a Kindle format book (*.azw).  Imported and failed to open.  Full details below.

I have not had this problem before.  Been great.

Any help much appreciated.

Regards

Stephen

Windows 11
Kindle book *.azw
Calibre  5.37.0
Updated Calibre.
Installed Alf - closed and re-opened Calibre

Added new Kindle Book - format *.xzw

Installed book - no title - just "B09C7HPQ7D EBOK"

Tried to open and got the following error details:

calibre, version 5.37.0
ERROR: Loading book failed: Failed to open the book at D:\Drive_Academic\00 Books\Books e-readers\Calibre Library\Unknown\B09C7HPQ7D EBOK (769)\B09C7HPQ7D EBOK - Unknown.azw. Click "Show details" for more info.

Failed to convert book: D:\Drive_Academic\00 Books\Books e-readers\Calibre Library\Unknown\B09C7HPQ7D EBOK (769)\B09C7HPQ7D EBOK - Unknown.azw with error:
DeDRM v7.2.1: Trying to decrypt B09C7HPQ7D EBOK - Unknown.azw
Using Library AlfCrypto DLL/DYLIB/SO
Using Library AlfCrypto DLL/DYLIB/SO
DeDRM v7.2.1: Failed to decrypt with error: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.1 seconds
searching for kinfoFiles in C:\Users\Subs\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\Subs\AppData\Local\Amazon\Kindle\storage\.kinf2018
Decrypted key file using IDString '2254096811' and UserName 'Subs'
DeDRM v7.2.1: Found 1 new key
DeDRM v7.2.1: Ultimately failed to decrypt after 2.9 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
InputFormatPlugin: MOBI Input running
on D:\Drive_Academic\00 Books\Books e-readers\Calibre Library\Unknown\B09C7HPQ7D EBOK (769)\B09C7HPQ7D EBOK - Unknown.azw
Failed to run pipe worker with command: from calibre.srv.render_book import viewer_main; viewer_main()
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 432, in get_file_type_metadata
  File "calibre\customize\builtins.py", line 257, in get_metadata
  File "calibre\ebooks\metadata\mobi.py", line 503, in get_metadata
  File "calibre\ebooks\mobi\reader\headers.py", line 293, in __init__
  File "calibre\ebooks\mobi\reader\headers.py", line 321, in identity
calibre.ebooks.mobi.MobiError: Unknown book type: b'\x02\xa2\x81\xc4\xde\x02\x9d\x8c'
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
  File "calibre_plugins.dedrm.k4mobidedrm", line 208, in GetDecryptedBook
calibre_plugins.dedrm.k4mobidedrm.DrmException: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 2.9 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Traceback (most recent call last):
  File "calibre\ebooks\conversion\plugins\mobi_input.py", line 27, in convert
  File "calibre\ebooks\mobi\reader\mobi6.py", line 89, in __init__
calibre.ebooks.mobi.reader.mobi6.KFXError: This is an Amazon KFX book. It cannot be processed. See https://www.mobileread.com/forums/showthread.php?t=283371 for information on how to handle KFX books.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\utils\ipc\worker.py", line 196, in main
  File "<string>", line 1, in <module>
  File "calibre\srv\render_book.py", line 823, in viewer_main
  File "calibre\srv\render_book.py", line 814, in render_for_viewer
  File "calibre\srv\render_book.py", line 791, in render
  File "calibre\ebooks\oeb\iterator\book.py", line 62, in extract_book
  File "calibre\customize\conversion.py", line 242, in __call__
  File "calibre\ebooks\conversion\plugins\mobi_input.py", line 33, in convert
  File "calibre\ebooks\mobi\reader\mobi6.py", line 89, in __init__
calibre.ebooks.mobi.reader.mobi6.KFXError: This is an Amazon KFX book. It cannot be processed. See https://www.mobileread.com/forums/showthread.php?t=283371 for information on how to handle KFX books.

**Comment by ElleKayEm** [comment]
As the log says, this is KFX format.  See the link given for dealing with or avoiding it:  https://www.mobileread.com/forums/showthread.php?t=283371

You may also want to update to noDRM's DeDRM:  https://github.com/noDRM/DeDRM_tools

**Reply by Chexote** [reply]
Thank you.  Appreciate the response.

Read the various materials.  
- used Kindle 1.17 to download books
- Updated to DeDRM 10.0.0 in Calibre

The book imported and can see metadata -- but cannot open.   Get the "protected by DRM" message.
Checked & retried everything.  Same result.

Have searched and read various other threads -- still no joy.

Cheers

**Reply by Chexote** [reply]
PS 
I have also tried older versions of Calibre and DeDRM without success.

Cheers

**Reply by ElleKayEm** [reply]
Let's see a new import log.

**Reply by Chexote** [reply]
Thank you.   I couldn't find an import log but stumbled onto debug mode.  Here is the log for that.  It still reflects the earlier version of Caibre I reinstalled.   I am getting the same result with the newest version.

calibre Debug log
calibre 4.23 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.22000-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22000')
Python 2.7.16+
Windows: ('10', '10.0.22000', 'SP0', u'Multiprocessor Free')
Interface language: en_GB
Successfully initialized third party plugins: DeDRM (6, 8, 0)
calibre 4.23 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.22000-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22000')
Python 2.7.16+
Windows: ('10', '10.0.22000', 'SP0', u'Multiprocessor Free')
Interface language: en_GB
Successfully initialized third party plugins: DeDRM (6, 8, 0)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 91.8418079096 x 91.6691729323
Using calibre Qt style: True
[0.00] Starting up...
[0.01] Showing splash screen...
[0.39] splash screen shown
[0.39] Initializing db...
[0.41] db initialized
[0.41] Constructing main UI...
[1.94] main UI initialized...
[1.94] Hiding splash screen
[17.25] splash screen hidden
[17.25] Started up in 17.25 seconds with 451 books
stdout+stderr from file dialog helper: ['', '']
piped data from file dialog helper: ['\xe2\x02H\x16\xe2\xaf\x8e\t\xd1\xfc1\x19\x19\x9c\xfb\xd734N\xd5\x81\xe3\x0e\xc0\xcc\x9b\xd8K\x0c\x99\xb2D', 'D:\\Drive_Academic\\00 Books\\Books e-readers\\Chess Books\\B003RCK1SO_EBOK.azw']
DeDRM v6.8.0: Trying to decrypt B003RCK1SO_EBOK.azw
Using Library AlfCrypto DLL/DYLIB/SO
MobiDeDrm v0.42.
Copyright © 2008-2017 The Dark Reverser, Apprentice Harper et al.
Decrypting Mobipocket 6 ebook: Elements of Positional Evaluation
Got DSN key from database default_key
Found 4 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v6.8.0: Failed to decrypt with error: No key found in 4 keys tried.
DeDRM v6.8.0: Looking for new default Kindle Key after 0.1 seconds
searching for kinfoFiles in C:\Users\Subs\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\Subs\AppData\Local\Amazon\Kindle\storage\.kinf2018
Decrypted key file using IDString '2254096811' and UserName '53756273'
DeDRM v6.8.0: Found 1 new key
MobiDeDrm v0.42.
Copyright © 2008-2017 The Dark Reverser, Apprentice Harper et al.
Decrypting Mobipocket 6 ebook: Elements of Positional Evaluation
Got DSN key from database default_key_1
Found 4 keys to try after 5.7 seconds
Crypto Type is: 2
DeDRM v6.8.0: Ultimately failed to decrypt after 5.7 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "site-packages\calibre\customize\ui.py", line 172, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 626, in run
  File "calibre_plugins.dedrm.__init__", line 576, in KindleMobiDecrypt
DeDRMError: DeDRM v6.8.0: Ultimately failed to decrypt after 5.7 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Added 0 books in 14.0 seconds
stdout+stderr from file dialog helper: ['', '']
piped data from file dialog helper: ["\xf3j\xde\x13y1\xaa'\x03\x03\xfa\x127`B@\xfd\xf96_\xa4\x10p1\xaf\tB\x91\xddn\x16\xab", 'D:\\Drive_Academic\\00 Books\\Books e-readers\\Chess Books\\B003RCK1SO_EBOK.azw']
DeDRM v6.8.0: Trying to decrypt B003RCK1SO_EBOK.azw
Using Library AlfCrypto DLL/DYLIB/SO
MobiDeDrm v0.42.
Copyright © 2008-2017 The Dark Reverser, Apprentice Harper et al.
Decrypting Mobipocket 6 ebook: Elements of Positional Evaluation
Got DSN key from database default_key
Found 4 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v6.8.0: Failed to decrypt with error: No key found in 4 keys tried.
DeDRM v6.8.0: Looking for new default Kindle Key after 0.1 seconds
searching for kinfoFiles in C:\Users\Subs\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\Subs\AppData\Local\Amazon\Kindle\storage\.kinf2018
Decrypted key file using IDString '2254096811' and UserName '53756273'
DeDRM v6.8.0: Found 1 new key
MobiDeDrm v0.42.
Copyright © 2008-2017 The Dark Reverser, Apprentice Harper et al.
Decrypting Mobipocket 6 ebook: Elements of Positional Evaluation
Got DSN key from database default_key_1
Found 4 keys to try after 6.1 seconds
Crypto Type is: 2
DeDRM v6.8.0: Ultimately failed to decrypt after 6.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "site-packages\calibre\customize\ui.py", line 172, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 626, in run
  File "calibre_plugins.dedrm.__init__", line 576, in KindleMobiDecrypt
DeDRMError: DeDRM v6.8.0: Ultimately failed to decrypt after 6.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Added Elements of Positional Evaluation to db in: 0.2
Added 1 books in 6.8 seconds

**Reply by ElleKayEm** [reply]
The version of Kindle for PC which DeDRM is retrieving the key from is newer then 1.17.  The books need to be downloaded with the Kindle app which is currently installed on the same machine and user as calibre with DeDRM.

**Comment by Chexote** [comment]
Thank you.  Some web pages said use Kindle to download so I did ... but on another computer.  I then transferred the file to my computer to importinto Calibre.   I will now try again on tje same computer.  I will let you know how I go.  Thank you for your help.Get Outlook for Android

**Reply by Chexote** [reply]
Thank you so much. really appreciate what you have done.   I now have the current version of Calibre on the computer I use to access the books and all is working well.

So I don't have to bother about unloading and loading the older version of Kindle (I do use Kindle as well) I have set up the current version of Calibre & DeDRM on a separate device together with the older version of Kindle.  On that PC when I have installed the books on Calibre I now just transfer the relevant Calibre book files to the device I am using and import.  Having set this up it is relatively quick.

From this, I have learnt about the new location of DeDRM and the need to have a copy of CAlibre and an older version of Kindle on the same computer.

Again, thank you very much I would have been lost without your help.

Cheers
